### PR TITLE
STAND REFACTOR [pls]

### PR DESCRIPTION
**Pull request by Militaires**
this new system allows for way more variety in stands, instead of just having TYPES there are abilities with different "Values" the higher the value the stronger the stand, but a stands total value can not exceed 10 so it maintains some form of balance all we gotta do is just give the abilities well-though values.

it's a datum of abilities added to a stand's abilities list, each datum ability gives it a new ability

basically way higher stando variety is to be expected, and anyone can add their own ability and stuff

- [x]  port all existing stands to new system
- [x]  create new more exciting stand types
- [x]  allow stands to use the new system
- [x]  implement a tool to summon the stand using the new system
- [x]  come back and add even more types

@changelingrain your blacklist system is in place but can you tell me which abilities are incompatible with what

## **GIANT DISCLAIMER: IM NOT SURE IF THE SPAWNER TOOL WORKS BECAUSE I CAN NOT PLAUSIBLY TEST IT IN MY LOCAL.**

### CODE REVIEW

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm [hunk 1]
```diff
@@ -1,100 +1,83 @@
...
+/datum/guardian_abilities/assassin/Destroy()
+	return ..()
 
-/mob/living/simple_animal/hostile/guardian/assassin/Life()
-	. = ..()
-	updatestealthalert()
-	if(loc == summoner && toggle)
-		ToggleMode(0)
+/datum/guardian_abilities/assassin/handle_stats()
+	guardian.has_mode = TRUE
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: Anything that has a mode isn't compatible with anything else that has a mode.

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm [hunk 2]
```diff
@@ -1,100 +1,83 @@
...
-		armour_penetration = initial(armour_penetration)
-		obj_damage = initial(obj_damage)
-		environment_smash = initial(environment_smash)
-		alpha = initial(alpha)
+		guardian.melee_damage_lower = initial(guardian.melee_damage_lower)
+		guardian.melee_damage_upper = initial(guardian.melee_damage_upper)
+		guardian.armour_penetration = initial(guardian.armour_penetration)
+		guardian.obj_damage = initial(guardian.obj_damage)
+		guardian.environment_smash = initial(guardian.environment_smash)
+		guardian.alpha = initial(guardian.alpha)
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: this needs to reset starts properly.

File: code/modules/mob/living/simple_animal/guardian/types/dextrous.dm [hunk 3]
```diff
@@ -1,102 +1,33 @@
...
-
-/mob/living/simple_animal/hostile/guardian/dextrous/getBeltSlot()
-	return slot_generic_dextrous_storage
-
-/mob/living/simple_animal/hostile/guardian/dextrous/proc/update_inv_internal_storage()
-	if(internal_storage && client && hud_used && hud_used.hud_shown)
-		internal_storage.screen_loc = ui_id
-		client.screen += internal_storage
+	guardian.drop_all_held_items()
+	return TRUE //lose items, then return
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/types/dextrous.dm]: This isn't compatible with protector guardians.

File: code/modules/mob/living/simple_animal/guardian/types/explosive.dm [hunk 4]
```diff
@@ -64,22 +65,22 @@
 
 /obj/guardian_bomb/proc/disable()
 	stored_obj.forceMove(get_turf(src))
-	to_chat(spawner, "<span class='danger'><B>Failure! Your trap didn't catch anyone this time.</span></B>")
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/types/explosive.dm]: unrevert all these to_chat()s.

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 5]
```diff
@@ -161,16 +235,34 @@ GLOBAL_LIST_EMPTY(parasites) //all currently existing/living guardians
...
+
+/mob/living/simple_animal/hostile/guardian/throw_impact(atom/A)
+	for(var/I in current_abilities)
+		var/datum/guardian_abilities/S = I
+		S.impact_act(A)
 
 /mob/living/simple_animal/hostile/guardian/proc/snapback()
+	for(var/I in current_abilities)
+		var/datum/guardian_abilities/S = I
+		S.snapback_act()
```
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: This needs to prevent the proc from continuing if the snapback isn't expecting it.

**Comment by ShadowDeath6** [comment]
Gotta thumbs down for that branch name

Good pr tho

**Comment by coiax** [comment]
Ask in coderbus how to use guest mode for testing

File: code/modules/mob/living/simple_animal/guardian/types/explosive.dm (hunk 4 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/explosive.dm]: woops

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 5 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: dont get what u mean

File: code/modules/mob/living/simple_animal/guardian/types/dextrous.dm (hunk 3 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/dextrous.dm]: why

Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/types/dextrous.dm]: protector guardians don't recall to their summoner, smart guy, they recall their summoner to them.

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 5 above)
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: like protector guardians, which should keep the proc from continuing.

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm (hunk 2 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: and how am i supposed to do that

if you're gonna tell me to do something 

either 1. tell me how to do it (preferred)
2. tell me WHY the current version does not work, so i can fix that specific mistake.

don't cop out if you're gonna review.

File: code/modules/mob/living/simple_animal/guardian/types/dextrous.dm (hunk 3 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/dextrous.dm]: yes but the guardians can still just pick up the thing and use it

whatever ill edit the snapback act.

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm (hunk 2 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: "you need to make this not bad"

"okay"

nice maintainer work

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 5 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: why cant you just make the snapback_act return 0 in the specific subtype you want it to stop in

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm (hunk 1 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: alright

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm (hunk 2 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: also what the hell does that mean "reset starts"

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 5 above)
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: because that doesn't keep this proc from continuing???

File: code/modules/mob/living/simple_animal/guardian/types/bounce.dm [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+/datum/guardian_abilities/bounce/ability_act(atom/movable/A)
+	var/atom/throw_target = get_edge_target_turf(A, guardian.dir)
+	A.throw_at(throw_target, bounce_distance, 14, guardian)
+
+/datum/guardian_abilities/bounce/boom_act(severity)
+	guardian.visible_message("<span class='danger'>The explosive force bounces off [guardian]'s rubbery surface!</span>")
+	for(var/mob/M in range(7,guardian))
+		if(M != user)
+			M.ex_act(severity) //reflect
+			M.visible_message("<span class='danger'>The explosive force bounces onto [M]!</span>")
```
Comment (XDTM) [inline code/modules/mob/living/simple_animal/guardian/types/bounce.dm]: if you drop a strong enough bomb on the stand you'll gib everyone onscreen

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 5 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: fixeddddddddddddddddddddd

File: code/modules/mob/living/simple_animal/guardian/types/bounce.dm (hunk 6 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/bounce.dm]: :^)

Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/types/bounce.dm]: helpfully, it will also gib the stand('s user)

Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/bounce.dm]: ^^^

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm (hunk 1 above)
Comment (Militaires) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: @ChangelingRain what if it's something that's kind of compatible like assassin and ranged?

**Comment by Militaires** [comment]
this needs a testicle-merge @korphaeron

File: code/modules/mob/living/simple_animal/guardian/types/assassin.dm (hunk 2 above)
Comment (optimumtact) [inline code/modules/mob/living/simple_animal/guardian/types/assassin.dm]: it's not our job to teach you how to code

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 7]
```diff
@@ -179,23 +273,43 @@ GLOBAL_LIST_EMPTY(parasites) //all currently existing/living guardians
...
+				msg += "<span class='warning'>It is holding \icon[internal_storage] [internal_storage.gender==PLURAL?"some":"a"] blood-stained [internal_storage.name] in its internal storage!</span>\n"
+			else
+				msg += "It is holding \icon[internal_storage] \a [internal_storage] in its internal storage.\n"
+		msg += "*---------*</span>"
+		to_chat(user,msg)
+
+
 
 /mob/living/simple_animal/hostile/guardian/death()
+	if(internal_storage && dextrous)
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: Why the dextrous check here ? Seems unrelated ?

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 8]
```diff
@@ -206,20 +320,27 @@ GLOBAL_LIST_EMPTY(parasites) //all currently existing/living guardians
 		hud_used.healths.maptext = "<div align='center' valign='middle' style='position:relative; top:0px; left:6px'><font color='#efeeef'>[resulthealth]%</font></div>"
 
 /mob/living/simple_animal/hostile/guardian/adjustHealth(amount, updating_health = TRUE, forced = FALSE) //The spirit is invincible, but passes on damage to the summoner
+	for(var/I in current_abilities)
+		var/datum/guardian_abilities/S = I
+		S.adjusthealth_act(amount, updating_health = TRUE, forced = FALSE)
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: You should probably be passing actual values here instead of defaults ?

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 9]
```diff
@@ -462,12 +654,22 @@ GLOBAL_LIST_EMPTY(parasites) //all currently existing/living guardians
...
 	var/theme = "magic"
-	var/mob_name = "Guardian Spirit"
+	var/mob_name = "guardian"
 	var/use_message = "<span class='holoparasite'>You shuffle the deck...</span>"
 	var/used_message = "<span class='holoparasite'>All the cards seem to be blank now.</span>"
 	var/failure_message = "<span class='holoparasitebold'>..And draw a card! It's...blank? Maybe you should try again later.</span>"
 	var/ling_failure = "<span class='holoparasitebold'>The deck refuses to respond to a souless creature such as you.</span>"
-	var/list/possible_guardians = list("Assassin", "Chaos", "Charger", "Explosive", "Lightning", "Protector", "Ranged", "Standard", "Support")
+
+	var/list/chosen_abilities = null
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: Why are these variables even here ? They are only used in creation proc as far as i can tell ?

**Comment by lzimann** [comment]
ping when you do the things and no conflicts
